build.subr

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+#
+# Shell script for building Phoenix-RTOS ports
+#
+# Common functions
+#
+# Copyright 2024 Phoenix Systems
+# Author: Daniel Sawka
+#
+
+# b_port_download(baseurl, filename, [url_filename])
+b_port_download() {
+	local baseurl="$1"
+	local filename="$2"
+	local url_filename="${3:-$filename}"
+
+	if [ ! -f "${PREFIX_PORT}/${filename}" ]; then
+		if ! wget "${baseurl}${url_filename}" -O "${PREFIX_PORT}/${filename}.part" --no-check-certificate; then
+			wget "${PORTS_MIRROR_BASEURL}${filename}" -O "${PREFIX_PORT}/${filename}.part" --no-check-certificate
+		fi
+
+		if [ -f "${PREFIX_PORT}/checksums.sha256" ]; then
+			while read -r line; do
+				local expected_checksum="${line%% *}"
+				local line_file="${line#*\*}" # shasum binary mode (*) expected
+
+				if [ -n "${expected_checksum}" ] && [ "${line_file}" = "${filename}" ]; then
+					local actual_checksum="$(shasum -b -a 256 "${PREFIX_PORT}/${filename}.part")"
+					actual_checksum="${actual_checksum%% *}"
+
+					if [ "${expected_checksum}" = "${actual_checksum}" ]; then
+						echo "Checksum OK for \"${filename}\""
+					else
+						b_die "Checksum INVALID for \"${filename}\": expected \"${expected_checksum}\", got \"${actual_checksum}\""
+					fi
+				fi
+			done < "${PREFIX_PORT}/checksums.sha256"
+		fi
+
+		mv -f "${PREFIX_PORT}/${filename}.part" "${PREFIX_PORT}/${filename}"
+	fi
+}
+
+# b_port_apply_patches(srcdir, [patch_subdir])
+b_port_apply_patches() {
+	local srcdir="$1"
+	local patch_subdir="${2:-.}"
+
+	local patch_dir="${PREFIX_PORT}/patches/${patch_subdir}"
```

Comment:
What is optional "patch_subdir" used for? Right now ports keep patches in main port dir or in "patches"/"patch" subdir.

Reply:
The general idea is to force patches to be put into _patches_ directory instead of the main port directory. Then, if one wants to, _patch_subdir_ may be used to group patches e.g. by version. That's what I do in lua port - I put patches in _patches/5.3.6_ directory, as they are not compatible with other versions, e.g. 5.4.7, and patch files would usually have the same names, e.g. _01-Makefile.patch_ for all versions (so they would have to e.g. be prefixed with versions). By default _patch_subdir_ is `.`, so this is an opt-in feature.